### PR TITLE
DAOS-11375 object: extend daos_obj_generate_oid_by_rf

### PR DESCRIPTION
**Pull request by wangshilong**
Now we have two different fault domains: rank and node.
daos_obj_generate_oid_by_rf() use default fault domain
which might be wrong. Eg, creating container(with rf_lvl:1)

pass in fault domain level to daos_obj_generate_oid_by_rf() which
is aware of fault domain.

Required-githooks: true
Signed-off-by: Wang Shilong <shilong.wang@intel.com>

### CODE REVIEW

File: src/include/daos/object.h [hunk 1]
```diff
@@ -443,6 +443,11 @@ int daos_obj_generate_oid_by_rf(daos_handle_t poh, uint64_t rf_factor,
 				daos_oclass_id_t cid, daos_oclass_hints_t hints,
 				uint32_t args);
 
+int daos_obj_generate_oid_by_rf2(daos_handle_t poh, uint64_t rf_factor,
```
Comment (mchaarawi) [inline src/include/daos/object.h:446]: this is not a public API, so you can just replace the old one with this new one.
no need to add API versioning here.

**Review by mchaarawi — CHANGES_REQUESTED** [review]

**Comment by mchaarawi** [comment]
can you please explain how this domain plays into the oclass selection?

**Comment by wangshilong** [comment]
> can you please explain how this domain plays into the oclass selection?

See dc_set_oclass() for example number of domains affects group size when object class is unknown, and we have check to
compare number of domains and group size, it will return invalid if  grp_size > domain_nr.

So it is quite important to get correct number of domains when generating OID.

File: src/common/cont_props.c [hunk 2]
```diff
@@ -224,7 +224,7 @@ daos_cont_prop2redunlvl(daos_prop_t *props)
 	struct daos_prop_entry *prop =
 		daos_prop_entry_get(props, DAOS_PROP_CO_REDUN_LVL);
 
-	return prop == NULL ? DAOS_PROP_CO_REDUN_RANK : (uint32_t)prop->dpe_val;
```
Comment (Nasf-Fan) [inline src/common/cont_props.c:227]: [suggestion] "#define DAOS_PROP_CO_REDUN_DEFAULT DAOS_PROP_CO_REDUN_NODE", replace the other DAOS_PROP_CO_REDUN_NODE usages with DAOS_PROP_CO_REDUN_DEFAULT. Then in the future, if we change the default domain as others (may be new), then we can just redefine DAOS_PROP_CO_REDUN_DEFAULT.

**Review by Nasf-Fan — APPROVED** [review]
> Introduce another helper daos_obj_generate_oid_by_rf2() which is aware of fault domain.

Please update the comment to match the latest patch if you have other chance to refresh it.

**Review by mchaarawi — APPROVED** [review]

**Review by Nasf-Fan — APPROVED** [review]

File: src/include/daos_obj.h [hunk 3]
```diff
@@ -113,6 +114,12 @@ daos_otype_t_is_valid(enum daos_otype_t type)
 	return type <= DAOS_OT_MAX;
 }
 
+static inline bool
+daos_pa_domain_is_valid(uint32_t pa_domain)
```
Comment (jolivier23) [inline src/include/daos_obj.h:118]: Should be a minor API version bump for adding a new API?

**Review by jolivier23 — CHANGES_REQUESTED** [review]
Need an answer to the question

File: src/include/daos_obj.h (hunk 3 above)
Comment (wangshilong) [reply src/include/daos_obj.h:118]: @mchaarawi do you think we should add a minor API version bump for this?

Comment (mchaarawi) [reply src/include/daos_obj.h:118]: i don't think this API will see much use, but i agree it's a good practice to bump the minor version when adding new APIs in any case.

**Review by mchaarawi — DISMISSED** [review]

**Review by Nasf-Fan — APPROVED** [review]

**Review by liuxuezhao — APPROVED** [review]
thanks @wangshilong for the fix.

**Comment by wangshilong** [comment]
@daos-stack/daos-gatekeeper pylint and Jira report failure are CI issue and not related to this PR, other tests passed.

**Comment by wangshilong** [comment]
@daos-stack/daos-gatekeeper PR modified api minor version which touched Sconcript, there are some other pylint errors triggerd there, but it is not related to this PR

File: SConstruct [hunk 4]
```diff
@@ -43,7 +43,7 @@ def get_version(env):
 
 
 API_VERSION_MAJOR = "2"
-API_VERSION_MINOR = "3"
+API_VERSION_MINOR = "4"
```
Comment (ashleypittman) [inline SConstruct:46]: https://github.com/daos-stack/daos/pull/9985 also changes this line and if the two PRs are merged then the change to the second one will be lost, so whichever PR lands seconds will need to rebase onto master after the first has landed and update this value to 5.

Comment (mchaarawi) [reply SConstruct:46]: since both updates are going on the same "release" and almost at the same time, there is no reason to bump this twice IMO

Comment (mchaarawi) [reply SConstruct:46]: anyway this is ready to land now, no?

Comment (kccain) [reply SConstruct:46]: hopefully https://github.com/daos-stack/daos/pull/9985 will not need a rebase if this lands first (and if we decide to keep the same minor version value). That PR just got rebased for another issue & retested. At the moment it does need a +1 before requesting landing.

Comment (wangshilong) [reply SConstruct]: Since this PR need be rebased due to conflict, i removed my API version bump and keep Ken's PR there.

**Review by Nasf-Fan — APPROVED** [review]

**Review by mchaarawi — APPROVED** [review]

**Comment by wangshilong** [comment]
@daos-stack/daos-gatekeeper PR is ready for landing now.

**Review by liuxuezhao — APPROVED** [review]

**Review by jolivier23 — APPROVED** [review]

**Pull request merged**